crypto/rand/drbg_lib.c

```diff
@@ -1031,19 +1031,85 @@ static int drbg_bytes(unsigned char *out, int count)
     return ret;
 }
 
+/*
+ * Calculates the minimum length of a full entropy buffer
+ * which is necessary to seed (i.e. instantiate) the DRBG
+ * successfully.
+ *
+ * NOTE: There is a copy of this function in drbgtest.c.
+ *       If you change anything here, you need to update
+ *       the copy accordingly.
+ */
+static size_t rand_drbg_seedlen(RAND_DRBG *drbg)
+{
+    /*
+     * If no os entropy source is available then RAND_seed(buffer, bufsize)
+     * is expected to succeed if and only if the buffer length satisfies
+     * the following requirements, which follow from the calculations
+     * in RAND_DRBG_instantiate().
+     */
+    size_t min_entropy = drbg->strength;
+    size_t min_entropylen = drbg->min_entropylen;
+
+    /*
+     * Extra entropy for the random nonce in the absence of a
+     * get_nonce callback, see comment in RAND_DRBG_instantiate().
+     */
+    if (drbg->min_noncelen > 0 && drbg->get_nonce == NULL) {
+        min_entropy += drbg->strength / 2;
+        min_entropylen += drbg->min_noncelen;
+    }
+
+    /*
+     * Convert entropy requirement from bits to bytes
+     * (dividing by 8 without rounding upwards, because
+     * all entropy requirements are divisible by 8).
+     */
+    min_entropy >>= 3;
+
+    /* Return a value that satisfies both requirements */
+    return min_entropy > min_entropylen ? min_entropy : min_entropylen;
+}
+
 /* Implements the default OpenSSL RAND_add() method */
 static int drbg_add(const void *buf, int num, double randomness)
 {
     int ret = 0;
     RAND_DRBG *drbg = RAND_DRBG_get0_master();
+    size_t buflen;
+    size_t seedlen = rand_drbg_seedlen(drbg);
 
     if (drbg == NULL)
         return 0;
 
     if (num < 0 || randomness < 0.0)
         return 0;
 
-    if (randomness > (double)RAND_DRBG_STRENGTH) {
+    buflen = (unsigned int)num;
+
+    if (buflen < seedlen || randomness < (double) seedlen) {
+#if defined(OPENSSL_RAND_SEED_NONE)
+        /*
+         * If no os entropy source is available, a reseeding will fail
+         * inevitably. So we use a trick to mix the buffer contents into
+         * the DRBG state without forcing a reseeding: we generate a
+         * dummy random byte, using the buffer content as additional data.
+         */
+        unsigned char dummy[1];
```

Comment:
Static analysers are going to report this as a problem.  Could we instead call `OPENSSL_rdtsc` and use the low byte of its result?  Failing that, a known constant value?
```
unsigned char dummy = OPENSSL_rdtsc();
return RAND_RDBG_generate(drbg, &dummy, sizeof(dummy), 0, buf, buflen);
```

Reply:
What would the static analyzer complain about? It is an uninitialized buffer of size 1, which is passed as out-parameter for being filled. Just like one would do it for a read() call. Or does the static analyzer infer from the fact that it is passed uninitialized the conclusion that it must be an out parameter and that the returned buffer contents are ignored?

I don't mind zero-initializing the buffer, but why does it have to be the return value of a fake call?

BTW: on second thought a buffer size of 1 seems a bit artificial to me. I am hovering between using sizeof(int)


```C
    unsigned int dummy = 0;

    return RAND_DRBG_generate(drbg, &dummy, sizeof(dummy), 0, buf, buflen);
```

or 16, the AES block size:

```C
    unsigned char dummy[16] = { 0 };

    return RAND_DRBG_generate(drbg, &dummy, sizeof(dummy), 0, buf, buflen);
```

What do you think?

Reply:
What would the static analyzer complain about? It is an uninitialized buffer of size 1, which is passed as out-parameter for being filled. Just like one would do it for a read() call. Or does the static analyzer infer from the fact that it is passed uninitialized the conclusion that it must be an out parameter and that the returned buffer contents are ignored?

I don't mind zero-initializing the buffer, but why does it have to be the return value of a fake call?

BTW: on second thought a buffer size of 1 seems a bit artificial to me. I am hovering between using sizeof(int)


```C
    unsigned int dummy = 0;

    return RAND_DRBG_generate(drbg, &dummy, sizeof(dummy), 0, buf, buflen);
```

or 16, the AES block size:

```C
    unsigned char dummy[16] = { 0 };

    return RAND_DRBG_generate(drbg, &dummy, sizeof(dummy), 0, buf, buflen);
```

What do you think?

Reply:
Update: Pauli and I agreed privately (since [GitHub was unreliable](https://blog.github.com/2018-10-21-october21-incident-report)) that the initialization is not an issue, because the dummy is used as out-parameter.

Reply:
I left the code unchanged.